core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -333,7 +353,11 @@ class BrokerLifecycleManager(val config: KafkaConfig,
   }
 
   private def sendBrokerHeartbeat(): Unit = {
-    val metadataOffset = _highestMetadataOffsetProvider()
+    val metadataOffset = if (readyToUnfence) {
```

Comment:
Why not use the published offset consistently in heartbeats?

Reply:
Following our current design in `BrokerServer`, when BrokerServer restarts, it will not start publishing metadata until it has caught up with the current high watermark, so the published offset will always be zero when starting.
This is necessary because when starting there may be some outdated metadata. for example, we created a topic and deleted it but the BrokerServer only read parts of the metadata log, the ReplicaManager will initialize some unnecessary replicas if it publishes at this point.

Reply:
I might still be missing something, but it seems ok if the published offset remains 0 initially and we send that in the heartbeat until the broker is ready to begin publishing. What do you think?

Reply:
This seems like the "Chicken and egg" problem. If we advertised 0 to the controller, the controller will always set isCaughtUp=false in HeartbeatResponse, the broker will never be ready to begin publishing.

The time sequence currently is: 
1. LifecycleManager repeatedly sends HeartBeatRequest
2. The controller set isCaughtUp=true if the advertised offset = high watermark
3. LifecycleManager marks initialCatchUpFuture is completed isCaughtUp=true 
4. The BrokerServer starts initial publishing after initialCatchUpFuture is completed

Reply:
Ok, I think I get it now. We don't begin publishing until the controller lets us know that we are caught up. That's the part I was missing. We might need to step back and reconsider this. Letting the reported offset in `BrokerHeartbeat` mean different things depending on the context seems undesirable. Does it give us any additional guarantees? Even after the controller marks the follower as "caught up," there is still a delay before the broker can publish the data. Additionally, would it be possible for the offset to go backwards when we switch to the published offset? It would be odd if we could end up going back to an offset which was not considered caught up.  What do you think?

Reply:
Currently, this offset is only used when starting the broker, first to ensure the broker has catchup so we can start initialization, second to ensure the broker has catchup so we can unfenced, if we use the same metadata offset here, it seems unnecessary to ensure twice, but It's indeed peculiar to have an offset going back, maybe adding a new field to the `HeartbeatRequest`, or just leave it as it is now?

Reply:
@dengziming Yeah, I was considering a second field as well. That would allow a startup transition looking something like this:

1. Broker catches up: fetch_offset=N, publish_offset=0
2. Controller returns is_caught_up=true, is_fenced=true
3. Broker begins publishing: fetch_offset=N, publish_offset=N
4. Controller returns is_caught_up=true, is_fenced=false

Basically the controller uses fetch_offset to tell whether the broker is caught up and it uses publish_offset to tell whether it should be unfenced. I guess the main advantage is that it gives us some more protection from exposing old metadata when brokers rejoin the cluster. Or at least we can put a time bound on the staleness of the exposed metadata. The value of this may be a bit limited though since we don't do self-fencing.

@cmccabe What do you think?

Reply:
This would need us to bump MetadataVersion so I will wait for KIP-778 first.